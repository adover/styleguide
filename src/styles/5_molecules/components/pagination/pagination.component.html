<ul class="pagination" aria-label="pagination">
  <li class="prev" [class.disabled]="previousDisabled()">
    <a
      aria-label="Previous"
      i18n-aria-label="@@pagination-previousLabel"
      href
      (click)="selectPage(currentPage - 1); $event.preventDefault()"
      [attr.tabindex]="hasPrevious() ? null : '-1'"
    >
      <cdx-svg-icon
        name="chevron-left"
        fill="dark"
        size="extra-small"
      ></cdx-svg-icon>
    </a>
  </li>
  <li
    *ngFor="let pageNumber of pages; let i = index"
    [class.selected]="pageNumber == currentPage"
    [class.disabled]="isEllipsis(pageNumber)"
  >
    <a aria-label="Ellipsis" *ngIf="isEllipsis(pageNumber)">...</a>
    <a
      aria-label="Page Number"
      *ngIf="!isEllipsis(pageNumber)"
      href
      (click)="selectPage(pageNumber); $event.preventDefault()"
    >
      {{ pageNumber }}
    </a>
  </li>
  <li class="next" [class.disabled]="nextDisabled()">
    <a
      aria-label="Next"
      i18n-aria-label="@@pagination-nextLabel"
      href
      (click)="selectPage(currentPage + 1); $event.preventDefault()"
      [attr.tabindex]="hasNext() ? null : '-1'"
    >
      <cdx-svg-icon
        name="chevron-right"
        fill="dark"
        size="extra-small"
      ></cdx-svg-icon>
    </a>
  </li>
</ul>
