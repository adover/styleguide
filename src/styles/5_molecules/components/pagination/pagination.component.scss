@import '../../../1_settings/colors';
@import '../../../2_tools/functions/svg.scss';
@import '../../../2_tools/mixins/list-reset.scss';
@import '../../../2_tools/mixins/display-flex.scss';
@import '../../../2_tools/mixins/pseudo.scss';
@import '../../../2_tools/mixins/center.scss';
@import '../../../4_atoms/list.scss';
@import '../../../4_atoms/link.scss';

:host {
	display: block;

	.pagination {
		@include list-reset();

		li {
			@include number-list-item();

			vertical-align: middle;

			.prev,
			.next {
				margin: 0;
			}

			a {
				@include display-flex($align: center, $justify: center);

				height: inherit;
				padding: 0 $spacing-1;
				text-align: center;
			}

			&.selected a {
				background-color: map-get($colors, parsley);
				color: white;

				a {
					@include link-dark();
				}
			}
		}
	}
}
