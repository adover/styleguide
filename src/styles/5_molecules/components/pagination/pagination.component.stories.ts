import { centered } from '@storybook/addon-centered/angular';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/angular';
import { StyleguideModule } from '../../../styleguide.module';

storiesOf('Molecules| Pagination', module)
	.addDecorator(withKnobs)
	.addDecorator(centered)
	.add('default', () => {
		const rotate = boolean('rotate', false);
		const ellipses = boolean('ellipses', true);
		const collectionSize = text('collectionSize', '100');
		const pagesToShow = text('maxSize', '3');
		const pageSize = text('pageSize', '10');
		return {
			moduleMetadata: {
				imports: [StyleguideModule],
			},
			template: `
        <cdx-pagination
          [currentPage]="1"
          [ellipses]="ellipses"
          [collectionSize]="collectionSize"
          [pageSize]="pageSize"
          [rotate]="rotate"
          [maxSize]="pagesToShow"
        >
        </cdx-pagination>
        <cdx-svg-definitions></cdx-svg-definitions>`,
			props: {
				ellipses,
				rotate,
				collectionSize,
				pagesToShow,
				pageSize,
			},
		};
	});
