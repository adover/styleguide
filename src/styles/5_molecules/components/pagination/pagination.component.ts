import { Component, EventEmitter, Input, OnInit, Output, SimpleChanges, OnChanges, HostBinding } from '@angular/core';

@Component({
	selector: 'cdx-pagination',
	templateUrl: './pagination.component.html',
	styleUrls: ['./pagination.component.scss'],
})
export class PaginationComponent implements OnInit, OnChanges {
	pageCount = 0;
	pages: number[] = [];
	@Input() ellipses = true;
	@Input() currentPage = 1;
	@Input() collectionSize: number;
	@Input() pageSize: number;
	@Input() maxSize = 3; // maximum pages to display
	@Input() rotate = false;

	@HostBinding('attr.role') role = 'navigation';

	@Output() pageChange: EventEmitter<number> = new EventEmitter();

	constructor() {}

	ngOnInit() {}

	ngOnChanges(changes: SimpleChanges): void {
		this.updatePages(this.currentPage);
	}

	applyEllipses(start: number, end: number) {
		if (this.ellipses) {
			if (end < this.pageCount) {
				if (end < this.pageCount - 1) {
					this.pages.push(-1);
				}
				this.pages.push(this.pageCount);
			}
		}
	}

	applyPagination(): [number, number] {
		if (this.currentPage === this.pageCount) {
			return [this.currentPage - this.maxSize - 1, this.currentPage];
		}
		const page = Math.ceil(this.currentPage / this.maxSize) - 1;
		const start = page * this.maxSize;
		const end = start + this.maxSize;

		return [start, end];
	}

	applyRotation(): [number, number] {
		let start = 0;
		let end = this.pageCount;
		const leftOffset = Math.floor(this.maxSize / 2);
		const rightOffset = this.maxSize % 2 === 0 ? leftOffset - 1 : leftOffset;

		if (this.currentPage <= leftOffset) {
			// very beginning, no rotation -> [0..maxSize]
			end = this.maxSize;
		} else if (this.pageCount - this.currentPage < leftOffset) {
			// very end, no rotation -> [len-maxSize..len]
			start = this.pageCount - this.maxSize;
		} else {
			// rotate
			start = this.currentPage - leftOffset - 1;
			end = this.currentPage + rightOffset;
		}

		return [start, end];
	}

	hasPrevious(): boolean {
		return this.currentPage > 1;
	}

	hasNext(): boolean {
		return this.currentPage < this.pageCount;
	}

	isEllipsis(pageNumber): boolean {
		return pageNumber === -1;
	}

	nextDisabled(): boolean {
		return !this.hasNext();
	}

	previousDisabled(): boolean {
		return !this.hasPrevious();
	}

	selectPage(pageNumber: number): void {
		if (!pageNumber || pageNumber > this.pageCount) return;
		this.updatePages(pageNumber);
	}

	setPageInRange(newPageNo) {
		const prevPageNo = this.currentPage;
		this.currentPage = newPageNo;

		if (this.currentPage !== prevPageNo && this.collectionSize) {
			this.pageChange.emit(this.currentPage);
		}
	}

	updatePages(newPage: number) {
		this.pageCount = Math.ceil(this.collectionSize / this.pageSize);

		if (!this.pageCount) {
			this.pageCount = 0;
		}
		this.pages.length = 0;
		for (let i = 1; i <= this.pageCount; i++) {
			this.pages.push(i);
		}
		this.setPageInRange(newPage);

		if (this.maxSize > 0 && this.pageCount > this.maxSize) {
			let start = 0;
			let end = this.pageCount;

			// either paginating or rotating page numbers
			if (this.rotate) {
				[start, end] = this.applyRotation();
			} else {
				[start, end] = this.applyPagination();
			}
			this.pages = this.pages.slice(start, end);
			this.applyEllipses(start, end);
		}
	}

	loadMore() {
		if (!this.hasNext()) return;
		this.currentPage += 1;
		this.pageChange.emit(this.currentPage);
	}
}
