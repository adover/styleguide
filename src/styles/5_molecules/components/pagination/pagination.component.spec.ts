import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { SvgIconComponent } from '@woolworthsnz/styleguide';
import { PaginationComponent } from './pagination.component';

describe('PaginationComponent', () => {
	let component: PaginationComponent;
	let fixture: ComponentFixture<PaginationComponent>;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			declarations: [PaginationComponent, SvgIconComponent],
			imports: [BrowserAnimationsModule],
		}).compileComponents();
	}));

	beforeEach(() => {
		fixture = TestBed.createComponent(PaginationComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('when rotation and ellipses are true', () => {
		component.pageSize = 10;
		component.collectionSize = 100;
		component.currentPage = 5;
		fixture.detectChanges();
		expect(fixture).toMatchSnapshot();
	});

	it('when rotation is false and ellipses is true', () => {
		component.pageSize = 10;
		component.collectionSize = 100;
		component.currentPage = 5;
		component.rotate = false;
		component.ellipses = true;
		expect(fixture).toMatchSnapshot();
	});

	it('when rotation and ellipses are false', () => {
		component.pageSize = 10;
		component.collectionSize = 100;
		component.currentPage = 5;
		component.rotate = false;
		component.ellipses = false;
		expect(fixture).toMatchSnapshot();
	});

	it('when rotation true and ellipses is false', () => {
		component.pageSize = 10;
		component.collectionSize = 100;
		component.currentPage = 5;
		component.rotate = true;
		component.ellipses = false;
		expect(fixture).toMatchSnapshot();
	});

	it('when rotation true and ellipses is false', () => {
		component.pageSize = 10;
		component.collectionSize = 100;
		component.currentPage = 5;
		component.rotate = true;
		component.ellipses = false;
		expect(fixture).toMatchSnapshot();
	});

	it('when click on next it should select next page', () => {
		jest.spyOn(component.pageChange, 'emit');
		jest.spyOn(component, 'selectPage');
		component.currentPage = 1;
		component.collectionSize = 100;
		component.pageSize = 10;
		component.pageCount = 10;
		fixture.debugElement.query(By.css('li.next a')).nativeElement.click();
		fixture.detectChanges();
		expect(component.selectPage).toHaveBeenCalledWith(2);
		expect(component.pageChange.emit).toHaveBeenCalledWith(2);
	});

	it('when click on prev it should select previous page', () => {
		jest.spyOn(component, 'selectPage');
		jest.spyOn(component.pageChange, 'emit');
		component.currentPage = 2;
		component.collectionSize = 100;
		component.pageSize = 10;
		component.pageCount = 10;
		fixture.debugElement.query(By.css('li.prev a')).nativeElement.click();
		fixture.detectChanges();
		expect(component.selectPage).toHaveBeenCalledWith(1);
		expect(component.pageChange.emit).toHaveBeenCalledWith(1);
	});

	it('when click on prev and cuurent page is 1 shouldnt do anything', () => {
		component.updatePages = jest.fn();
		component.currentPage = 1;
		fixture.detectChanges();
		fixture.debugElement.query(By.css('li.prev a')).nativeElement.click();
		expect(component.updatePages).not.toHaveBeenCalled();
	});

	it('when click on next and cuurent page is last page shouldnt do anything', () => {
		component.updatePages = jest.fn();
		component.pageCount = 10;
		component.currentPage = 10;

		fixture.detectChanges();
		fixture.debugElement.query(By.css('li.next a')).nativeElement.click();
		expect(component.updatePages).not.toHaveBeenCalled();
	});
});
