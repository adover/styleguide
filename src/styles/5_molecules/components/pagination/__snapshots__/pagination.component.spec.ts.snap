// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationComponent when rotation and ellipses are false 1`] = `
<cdx-pagination
  __ngContext__={[Function LRootView]}
  collectionSize={[Function Number]}
  currentPage={[Function Number]}
  ellipses="false"
  maxSize={[Function Number]}
  pageChange={[Function EventEmitter]}
  pageCount="0"
  pageSize={[Function Number]}
  pages={[Function Array]}
  role={[Function String]}
  rotate="false"
>
  <ul
    aria-label="pagination"
    class="pagination"
  >
    <li
      class="prev disabled"
    >
      <a
        aria-label="Previous"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-left"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-left"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
    
    <li
      class="next disabled"
    >
      <a
        aria-label="Next"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-right"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-right"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
  </ul>
</cdx-pagination>
`;

exports[`PaginationComponent when rotation and ellipses are true 1`] = `
<cdx-pagination
  __ngContext__={[Function LRootView]}
  collectionSize={[Function Number]}
  currentPage={[Function Number]}
  ellipses={[Function Boolean]}
  maxSize={[Function Number]}
  pageChange={[Function EventEmitter]}
  pageCount="0"
  pageSize={[Function Number]}
  pages={[Function Array]}
  role={[Function String]}
  rotate="false"
>
  <ul
    aria-label="pagination"
    class="pagination"
  >
    <li
      class="prev"
    >
      <a
        aria-label="Previous"
        href=""
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-left"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-left"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
    
    <li
      class="next disabled"
    >
      <a
        aria-label="Next"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-right"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-right"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
  </ul>
</cdx-pagination>
`;

exports[`PaginationComponent when rotation is false and ellipses is true 1`] = `
<cdx-pagination
  __ngContext__={[Function LRootView]}
  collectionSize={[Function Number]}
  currentPage={[Function Number]}
  ellipses={[Function Boolean]}
  maxSize={[Function Number]}
  pageChange={[Function EventEmitter]}
  pageCount="0"
  pageSize={[Function Number]}
  pages={[Function Array]}
  role={[Function String]}
  rotate="false"
>
  <ul
    aria-label="pagination"
    class="pagination"
  >
    <li
      class="prev disabled"
    >
      <a
        aria-label="Previous"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-left"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-left"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
    
    <li
      class="next disabled"
    >
      <a
        aria-label="Next"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-right"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-right"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
  </ul>
</cdx-pagination>
`;

exports[`PaginationComponent when rotation true and ellipses is false 1`] = `
<cdx-pagination
  __ngContext__={[Function LRootView]}
  collectionSize={[Function Number]}
  currentPage={[Function Number]}
  ellipses="false"
  maxSize={[Function Number]}
  pageChange={[Function EventEmitter]}
  pageCount="0"
  pageSize={[Function Number]}
  pages={[Function Array]}
  role={[Function String]}
  rotate={[Function Boolean]}
>
  <ul
    aria-label="pagination"
    class="pagination"
  >
    <li
      class="prev disabled"
    >
      <a
        aria-label="Previous"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-left"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-left"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
    
    <li
      class="next disabled"
    >
      <a
        aria-label="Next"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-right"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-right"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
  </ul>
</cdx-pagination>
`;

exports[`PaginationComponent when rotation true and ellipses is false 2`] = `
<cdx-pagination
  __ngContext__={[Function LRootView]}
  collectionSize={[Function Number]}
  currentPage={[Function Number]}
  ellipses="false"
  maxSize={[Function Number]}
  pageChange={[Function EventEmitter]}
  pageCount="0"
  pageSize={[Function Number]}
  pages={[Function Array]}
  role={[Function String]}
  rotate={[Function Boolean]}
>
  <ul
    aria-label="pagination"
    class="pagination"
  >
    <li
      class="prev disabled"
    >
      <a
        aria-label="Previous"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-left"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-left"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
    
    <li
      class="next disabled"
    >
      <a
        aria-label="Next"
        href=""
        tabindex="-1"
      >
        <cdx-svg-icon
          display="block"
          fill="dark"
          name="chevron-right"
          size="extra-small"
        >
          <svg
            version="1.1"
            viewbox="0 0 32 32"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <use
              xlink:href="#chevron-right"
            />
          </svg>
        </cdx-svg-icon>
      </a>
    </li>
  </ul>
</cdx-pagination>
`;
